File: migrate/src/java/com/tacitknowledge/util/migration/jdbc/SqlScriptMigrationTaskSource.java
/* 
 * Copyright 2007 Tacit Knowledge LLC
 * 
 * Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
 * you may not use this file except in compliance with the License. You may
 * obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tacitknowledge.util.migration.jdbc;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import com.tacitknowledge.util.discovery.ClassDiscoveryUtil;
import com.tacitknowledge.util.migration.MigrationException;
import com.tacitknowledge.util.migration.MigrationTaskSource;

/**
 * Search a package (directory) for SQL scripts that a specific pattern and
 * returns corresponding <code>SqlScriptMigrationTasks</code>. The name of
 * each script must follow the pattern of &quot;patch(\d+)(_.+)?\.sql&quot;.
 * 
 * @author Scott Askew (scott@tacitknowledge.com)
 */
public class SqlScriptMigrationTaskSource implements MigrationTaskSource
{
    /** Class logger */
    private static Log log = LogFactory
	    .getLog(SqlScriptMigrationTaskSource.class);

    /**
     * The regular expression used to match SQL patch files.
     */
     private static final String SQL_PATCH_REGEX = "^patch(\\d+)_(.+)\\.sql";

    /**
     * The regular expression used to match SQL rollback files
     */
    private static final String SQL_ROLLBACK_REGEX = "^patch(\\d+)-rollback_(.+)\\.sql";

    /** {@inheritDoc} */
    public List getMigrationTasks(String packageName) throws MigrationException
    {
	String path = packageName.replace('.', '/');

	String[] upScripts = getSqlScripts(path,SQL_PATCH_REGEX);
	String[] downScripts = getSqlScripts(path, SQL_ROLLBACK_REGEX);

	return createMigrationScripts(upScripts, downScripts);
    }

    /**
     * Returns the SQL scripts which exist in the specified patch and match the specified regex.
     * @param path the Path to search
     * @param regex the regex which the scripts should match 
     * @return a String array of script names.
     */
    private String[] getSqlScripts(String path, String regex)
    {
	String[] scripts = ClassDiscoveryUtil.getResources(path, regex);
	if (log.isDebugEnabled())
	{
	    log.debug("Found " + scripts.length + " patches in path: " + path);
	    for (int i = 0; i < scripts.length; i++)
	    {
		log.debug(" -- \"" + scripts[i] + "\"");
	    }
	}
	return scripts;
    }

    /**
     * Creates a list of <code>SqlScriptMigrationTask</code>s based on the
     * array of SQL scripts.
     * 
     * @param scripts
     *                the classpath-relative array of SQL migration scripts
     * @return a list of <code>SqlScriptMigrationTask</code>s based on the
     *         array of SQL scripts
     * @throws MigrationException
     *                 if a SqlScriptMigrationTask could no be created
     */
    private List createMigrationScripts(String[] upScripts, String[] downScripts)
	    throws MigrationException
    {
	Pattern upFileNamePattern = Pattern.compile(SQL_PATCH_REGEX);
	Pattern downFileNamePattern = Pattern.compile(SQL_ROLLBACK_REGEX);

	List tasks = new ArrayList();
	for (int i = 0; i < upScripts.length; i++)
	{
	    String script = upScripts[i];

	    if (script != null)
	    {
		// get the file name
		File scriptFile = new File(script);
		String scriptFileName = scriptFile.getName();

		// Get the version out of the script name
		int order = getOrder(upFileNamePattern, script, scriptFileName);

		// get the down script which matches this patch order
		String downScript = getMatchingDownScript(downScripts, order,
			downFileNamePattern);

		// read the scripts
		String upSql = readSql(getInputStream(script));
		String downSql = "";

		// if the down script is not the empty string, then try to read
		// it.
		if (!"".equals(downScript))
		    downSql = readSql(getInputStream(downScript));

		// create a new task
		SqlScriptMigrationTask task = new SqlScriptMigrationTask(
			scriptFileName, order, upSql, downSql);
		task.setName(scriptFileName);

		// add the task to the list of tasks
		tasks.add(task);
	    }
	}
	return tasks;
    }

    /**
     * Returns the filename of the downscript
     * 
     * @param downScripts
     * @param order
     * @return
     */
    private String getMatchingDownScript(String[] downScripts, int order,
	    Pattern fileNamePattern) throws MigrationException
    {
	boolean isScriptFound = false;
	String script = "";
	for (int i = 0; i < downScripts.length && !isScriptFound; i++)
	{
	    File scriptFile = new File(downScripts[i]);
	    String scriptFileName = scriptFile.getName();
	    int downScriptOrder = getOrder(fileNamePattern, downScripts[i],
		    scriptFileName);

	    if (downScriptOrder == order)
	    {
		script = downScripts[i];
		isScriptFound = true;
	    }
	}

	if (!isScriptFound)
	{
	    log.info("There was no rollback script for patch level: " + order);
	}
	return script;
    }

    /**
     * Returns an input stream that points to the script name
     * 
     * @param scriptName the name of the script to create an InputStream
     * @return an InputStream returns an InputStream based upon the scriptName
     */
    private InputStream getInputStream(String scriptName)
    {
	scriptName = scriptName.replace('\\', '/');
	log.debug("Examining possible SQL patch file \"" + scriptName + "\"");
	return Thread.currentThread().getContextClassLoader()
		.getResourceAsStream(scriptName);
    }

    /**
     * Reads the file contents into a String object.
     * 
     * @param is
     *                the <code>InputStream</code>
     * @return a <code>String</code> with the contents of the InputStream
     * @throws MigrationException
     *                 if there's an error reading in the contents
     */
    private String readSql(InputStream is) throws MigrationException
    {
	StringBuffer sqlBuffer = new StringBuffer();
	BufferedReader buf = new BufferedReader(new InputStreamReader(is));

	try
	{
	    String line = buf.readLine();
	    while (line != null)
	    {
		sqlBuffer.append(line).append("\n");
		line = buf.readLine();
	    }
	} catch (IOException ioe)
	{
	    throw new MigrationException(
		    "There was an error reading in a script", ioe);

	} finally
	{
	    try
	    {
		is.close();
	    } catch (IOException ioe)
	    {
		log.error("Could not close input stream", ioe);
	    }
	}
	return sqlBuffer.toString();
    }

    /**
     * Returns the order for the file.
     * 
     * @param p
     *                a Pattern defining the file name pattern
     * @param script
     *                the Script
     * @param scriptFileName
     *                the name of the file
     * @return an int indicating the order
     * @throws MigrationException
     *                 in case the file name is invalid
     */
    private int getOrder(Pattern p, String script, String scriptFileName)
	    throws MigrationException
    {
	Matcher matcher = p.matcher(scriptFileName);
	if (!matcher.matches() || matcher.groupCount() != 2)
	{
	    throw new MigrationException("Invalid SQL script name: " + script);
	}
	int order = Integer.parseInt(matcher.group(1));
	return order;
    }
}

File: migrate/dotnet/src/com/tacitknowledge/util/migration/ADO/SqlScriptMigrationTaskSource.cs
/* Copyright 2006 Tacit Knowledge LLC
* 
* Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
* you may not use this file except in compliance with the License. You may
* obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
#region Imports
using System;
//UPGRADE_TODO: The type 'java.util.regex.Matcher' could not be found. If it was not included in the conversion, there may be compiler issues. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1262'"
//using Matcher = java.util.regex.Matcher;
//UPGRADE_TODO: The type 'java.util.regex.Pattern' could not be found. If it was not included in the conversion, there may be compiler issues. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1262'"
//using Pattern = java.util.regex.Pattern;
using log4net;
using log4net.Config;

using MigrationException = com.tacitknowledge.util.migration.MigrationException;
using MigrationTaskSource = com.tacitknowledge.util.migration.MigrationTaskSource;
#endregion
namespace com.tacitknowledge.util.migration.ado
{
	
	/// <summary> Search a package (directory) for SQL scripts that a specific pattern 
	/// and returns corresponding <code>SqlScriptMigrationTasks</code>.  The name
	/// of each script must follow the pattern of &quot;patch(\d+)(_.+)?\.sql&quot;.
	/// 
	/// </summary>
	/// <author>   Scott Askew (scott@tacitknowledge.com)
	/// </author>
	public class SqlScriptMigrationTaskSource : MigrationTaskSource
    {
        #region Methods
        /// <summary>Class logger </summary>
		//UPGRADE_NOTE: The initialization of  'log' was moved to static method 'com.tacitknowledge.util.migration.ado.SqlScriptMigrationTaskSource'. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1005'"
		private static ILog log;
		
		/// <summary> The regular expression used to match SQL patch files.</summary>
		private const System.String SQL_PATCH_REGEX = "^patch(\\d+)(_.+)?\\.sql";
		
		/// <seealso cref="MigrationTaskSource.getMigrationTasks(String)">
		/// </seealso>
		public virtual System.Collections.IList getMigrationTasks(System.String packageName)
		{
			System.String path = packageName.Replace('.', '/');
			System.String[] scripts = ClassDiscoveryUtil.getResources(path, SQL_PATCH_REGEX);
			if (log.IsDebugEnabled())
			{
				log.Debug("Found " + scripts.Length + " patches in path: " + path);
				for (int i = 0; i < scripts.Length; i++)
				{
					log.Debug(" -- \"" + scripts[i] + "\"");
				}
			}
			return createMigrationScripts(scripts);
		}
		
		/// <summary> Creates a list of <code>SqlScriptMigrationTask</code>s based on the array
		/// of SQL scripts.
		/// 
		/// </summary>
		/// <param name="scripts">the classpath-relative array of SQL migration scripts
		/// </param>
		/// <returns> a list of <code>SqlScriptMigrationTask</code>s based on the array
		/// of SQL scripts
		/// </returns>
		/// <throws>  MigrationException if a SqlScriptMigrationTask could no be created </throws>
		private System.Collections.IList createMigrationScripts(System.String[] scripts)
		{
			Pattern p = Pattern.compile(SQL_PATCH_REGEX);
			System.Collections.IList tasks = new System.Collections.ArrayList();
			for (int i = 0; i < scripts.Length; i++)
			{
				System.String script = scripts[i];
				script = script.Replace('\\', '/');
				log.Debug("Examining possible SQL patch file \"" + script + "\"");
				//UPGRADE_ISSUE: Method 'java.lang.Class.getResourceAsStream' was not converted. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1000_javalangClassgetResourceAsStream_javalangString'"
				System.IO.Stream is_Renamed = GetType().getResourceAsStream("/" + script);
				
				if (is_Renamed == null)
				{
					log.warn("Could not open input stream for file \"/" + script + "\"");
				}
				else
				{
					System.IO.FileInfo scriptFile = new System.IO.FileInfo(script);
					System.String scriptFileName = scriptFile.Name;
					try
					{
						// Get the version out of the script name
						Matcher matcher = p.matcher(scriptFileName);
						if (!matcher.matches() || matcher.groupCount() != 2)
						{
							throw new MigrationException("Invalid SQL script name: " + script);
						}
						int order = Integer.parseInt(matcher.group(1));
						
						// We should send in the script file location so
						// it doesn't have to buffer the whole thing into RAM
						SqlScriptMigrationTask task = new SqlScriptMigrationTask(scriptFileName, order, is_Renamed);
						
						// Free the resource
						is_Renamed.Close();
						task.setName(scriptFileName);
						tasks.Add(task);
					}
					catch (System.IO.IOException e)
					{
						throw new MigrationException("Error reading script " + script, e);
					}
				}
			}
			return tasks;
		}
		static SqlScriptMigrationTaskSource()
		{
			log = LogManager.GetLogger(typeof(SqlScriptMigrationTaskSource));
		}
	}
#endregion
}